pkg/fqdn/regexpmap/regexp_map.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package regexpmap
+
+import (
+	"regexp"
+	"sort"
+)
+
+// lookupValueSet is a utility type. It is intended as a set of strings
+// inserted into a RegexpMap
+type lookupValueSet map[string]struct{}
+
+// RegexpMap is a map-like type that allows lookups to match regexp keys. These
+// keys are managed internally as strings, and are uniqued by this
+// representation (and not recompiled on repeat inserts).
+// Stored values are strings and are returned as-is, with the exception that
+// repeat inserts of the same value are de-duped.
+// Note: RegexpMap is not thread-safe and managing concurrent access is the
+// responsibility of the callers.
+type RegexpMap struct {
+	// lookups maps the original key string to the values added with it in .Add.
+	// These values are returned when a "key" (the matching regex in rules)
+	// matches.
+	lookups map[string]lookupValueSet
+	// rules maps the original key strong to the compiled regex for it. A .Lookup
+	// call will iterate over all the values in this map.
+	rules map[string]*regexp.Regexp
+}
+
+// NewRegexpMap returns an initialized RegexpMap
+func NewRegexpMap() *RegexpMap {
+	return &RegexpMap{
+		lookups: make(map[string]lookupValueSet),
+		rules:   make(map[string]*regexp.Regexp),
+	}
+}
+
+// Add associates lookupValue as a return value for reStr on lookup
+// Repeated calls with the same reStr will not recompile the regexp, but will
+// store all lookupValues in a set
+func (m *RegexpMap) Add(reStr string, lookupValue string) error {
+	// if this is the first Add of reStr, compile the regexp and setup the
+	// lookupValue set
+	if _, exists := m.rules[reStr]; !exists {
+		rule, err := regexp.Compile(reStr)
+		if err != nil {
+			return err
+		}
+		m.rules[reStr] = rule
+		m.lookups[reStr] = lookupValueSet{}
+	}
+
+	// add the lookupValue to the set for reStr
+	m.lookups[reStr][lookupValue] = struct{}{}
+
+	return nil
+}
+
+// LookupValues returns all lookupValues, inserted via Add, where the reStr
+// matches lookupKey
+func (m *RegexpMap) LookupValues(lookupKey string) (lookupValues []string) {
+	for reStr, rule := range m.rules {
+		if !rule.MatchString(lookupKey) {
```

Comment:
We may see an issue with contention on the regexes themselves (which have an internal lock). In newer Gos map iteration is random so I'm hopeful that small numbers of concurrent readers would see the regexes in different orders and never contend on the lock.